app/Observers/PaymentObserver.php

```diff
@@ -27,5 +29,22 @@ public function saved(Payment $payment): void
         ) {
             SendDuesPaymentReminder::dispatch($payment->payable->user);
         }
+
+        // this is pretty cursed but i don't have a better idea on guaranteeing exactly one receipt email from
+        // ~four save events
+        Cache::lock('send_payment_receipt_'.$payment->id, 5)->get(static function () use ($payment): void {
```

Comment:
The documentation about this function is lacking details and I found myself seeking out answers in the original source code.  The biggest issue is what the 5 means, I would suggest extracting it into a variable and adding a comment that it'll wait up to 5 seconds to obtain the lock ([if I understand correctly](https://github.com/laravel/framework/blob/9.x/src/Illuminate/Cache/Lock.php#L26)).

Reply:
That was also my understanding. I'll add a comment.